### ISSUE TITLE
I trained connect4 game 4624102 count. used 9 days. but muzero is stupid.

### ISSUE THREAD

**Issue by zfavourite99** [werner-duvaud/muzero-general#13]
I have changed some params:

        ### Game
        self.observation_shape = 6 * 7  # Dimensions of the game observation
        self.action_space = [i for i in range(7)]  # Fixed list of all possible actions
        self.players = [i for i in range(2)]  # List of players
        self.stacked_observations = 2  # Number of previous observation to add to the current observation

        ### Self-Play
        # 自我播放以供重播缓冲区使用的并发线程数
        # Number of simultaneous threads self-playing to feed the replay buffer
        self.num_actors = 8  
        # 如果游戏未完成之前的最大移动次数 
        # Maximum number of moves if game is not finished before
        self.max_moves = 50  
        # 自我模拟的未来动作数
        # Number of futur moves self-simulated
        self.num_simulations = 30  
        # 奖励按时间顺序折扣
        # Chronological discount of the reward
        self.discount = 0.997  
        # 每次玩游戏后需要等待的秒数，以调整自我玩耍/训练比率，以免过度/不足
        # Number of seconds to wait after each played game to adjust the self play / training ratio to avoid over/underfitting
        self.self_play_delay = 0 

        # Root prior exploration noise
        self.root_dirichlet_alpha = 0.25
        self.root_exploration_fraction = 0.25

        # UCB formula
        self.pb_c_base = 19652
        self.pb_c_init = 1.25

        ### Network
        self.encoding_size = 32
        self.hidden_layers = [64]

        # 价值和奖励按比例缩放（几乎为sqrt）并编码在范围为-support_size到support_size的向量上
        # Value and reward are scaled (with almost sqrt) and encoded on a vector with a range of -support_size to support_size
        self.support_size = 10  

        ### Training
        # 存储模型权重的路径
        # Path to store the model weights
        self.results_path = "./pretrained" 

        # 培训步骤总数（即，权重根据批次进行更新）
        # Total number of training steps (ie weights update according to a batch)
        self.training_steps = 1000000  

        # 每个训练步骤要训练的游戏零件数
        # Number of parts of games to train on at each training step
        self.batch_size = 512  

        # 每个批次元素要保留的游戏移动次数
        # Number of game moves to keep for every batch element
        self.num_unroll_steps = 10  

        # 在使用模型进行自弹奏之前的训练步骤数
        # Number of training steps before using the model for sef-playing
        self.checkpoint_interval = 10  

        # 保留在重播缓冲区中的自玩游戏数
        # Number of self-play games to keep in the replay buffer
        self.window_size = 1000  

        # 计算目标值时要考虑的未来步骤数
        # Number of steps in the futur to take into account for calculating the target value
        self.td_steps = 10  

        # 每次训练后需要等待的秒数，以调整自我发挥/训练比率，以避免过度/不足
        # Number of seconds to wait after each training to adjust the self play / training ratio to avoid over/underfitting
        self.training_delay = 0 

        # Train on GPU if available
        self.training_device = "cuda" if torch.cuda.is_available() else "cpu"  

        self.weight_decay = 1e-4  # L2 weights regularization
        self.momentum = 0.9

        # Exponential learning rate schedule
        self.lr_init = 0.05  # Initial learning rate
        self.lr_decay_rate = 1
        self.lr_decay_steps = 10000

        ### Test
        self.test_episodes = 2  # Number of game played to evaluate the network

![Screenshot from 2020-02-27 11-21-37](https://user-images.githubusercontent.com/2630290/75409739-8dd6a200-5954-11ea-8293-3c01f335ee7e.png)

error in training. maybe because I install gnome-tewaks-tool in training. Error is after i install this software.

this is i play with muzero (allways muzero first):

![muzero-test](https://user-images.githubusercontent.com/2630290/75409972-47357780-5955-11ea-9bcc-b648ee2e13dc.gif)

My environment is:
Ubuntu 18.04
i7-7700   16G
RTX 2080   8G

how to train a smart muzero for connect4?

**Comment by werner-duvaud** [comment]
Hi,

It's great to see the interest you have in the project.

The project is under development, we are still testing and optimizing Muzero for multiplayer board games.

The version you are using uses a fully connected network, we have since released a mode with a residual network which is much more suitable for board games.

Regarding the hyperparameters, we are conducting tests to select the best ones but it takes a long time.
To start, the td_steps is not good, it must be greater than the number of moves necessary to end a game because the reward is only given at the end of the game. Without this, the network cannot learn whether certain moves are better than others.

About the error that you had, it is related to a consumption of too much RAM memory. MuZero is coded to exploit almost 100% of the available resources. It is very likely that the ram memory was almost full (Ray automatically adjusts this).  Installing something in parallel had to temporarily increase the RAM memory and Ray stopped the training to avoid freezing the computer due to a full RAM memory.

Do not hesitate to keep me informed of your results.
